tutorbackup/templates/backup/build/backup/backup_services.py

```diff
@@ -0,0 +1,186 @@
+    outdir = MONGODB_DUMPDIR
+
+    logger.info(f"Dumping MongoDB databases on {host}:{port} to {outdir}")
+    cmd = ("mongodump "
+           f"--out={outdir} "
+           f"--host={host} "
+           f"--port={port}")
+    try:
+        cmd += (f" --username={ENV['MONGODB_USERNAME']} "
+                f"--password={ENV['MONGODB_PASSWORD']}")
+    except KeyError:
+        pass
+
+    check_call(cmd,
+               shell=True,
+               stdout=sys.stdout,
+               stderr=sys.stderr)
+
+    total_size = get_size(outdir)
+    logger.info(f"Complete. {outdir} total size {total_size} bytes.")
+
+
+def caddydump():
+    outdir = CADDY_DUMPDIR
+    logger.info(f"Copying Caddy data to {outdir}")
+    shutil.copytree('caddy', outdir)
+
+    total_size = get_size(outdir)
+    logger.info(f"Complete. {outdir} total size {total_size} bytes.")
+
+
+def archive():
+    paths = [MYSQL_DUMPFILE, MONGODB_DUMPDIR]
+    if PERFORM_CADDY_BACKUP:
+        paths.append(CADDY_DUMPDIR)
+
+    outfile = TARFILE
+
+    logger.info(f"Creating archive {outfile}")
+    with tarfile.open(outfile, "w:xz") as tar:
+        for item in paths:
+            tar.add(item)
+
+    size = os.path.getsize(outfile)
+    logger.info(f"Complete. {outfile} is {size} bytes.")
+
+
+def upload_to_s3():
+    import hashlib
+    from s3_client import S3_CLIENT
+
+    bucket = ENV['S3_BUCKET_NAME']
+    file_name = TARFILE
+
+    logger.info(f"Uploading {file_name} to S3 bucket {bucket}")
+    try:
+        S3_CLIENT.upload_file(
+            file_name,
+            bucket,
+            os.path.basename(file_name),
+        )
+        logger.info(f"Uploaded {file_name} to {bucket}.")
+
+        logger.info("Checking uploaded file's integrity ...")
+        obj_metadata = S3_CLIENT.head_object(
+            Bucket=bucket,
+            Key=os.path.basename(file_name),
+        )
+
+        received_checksum = obj_metadata['ETag'].strip('"')
+        calculated_checksum = hashlib.md5(
+            open(file_name, 'rb').read()).hexdigest()
+        version_id = obj_metadata['VersionId']
+        size = obj_metadata['ContentLength']
+
+        if received_checksum == calculated_checksum:
+            logger.info("File integrity verified.")
+            logger.info(f"Version ID: '{version_id}'")
+            logger.info(f"Size: {size} bytes")
+            logger.info(f"Checksum: '{received_checksum}'")
+        else:
+            S3_CLIENT.delete_object(
+                Bucket=bucket,
+                Key=os.path.basename(file_name),
+                VersionId=version_id,
+            )
+            raise ClientError(
+                msg=f"File integrity could not be verified. "
+                    f"Deleted the uploaded version "
+                    f"(VersionId: {version_id})."
+            )
+
+    except ClientError as e:
+        logger.exception(e, exc_info=True)
+
+
+@click.command()
+@click.option('--upload', is_flag=True, help="Upload to S3")
```

Comment:
I'd suggest to add

```python
@click.option("--exclude", multiple=True, default=[], help="Services to exclude from backup")
```

(Maybe you can also add something like `type=click.Choice(['caddy', 'mysql', 'mongodb']`, but I don't know off the top of my head how `click.Choice` works with `multiple=True`.)

And then instead of

```python
if PERFORM_CADDY_BACKUP:
    caddydump()
```

use 

```python
if 'caddy' not in exclude:
    caddydump()
```

What do you think?

Reply:
I'll make a follow up PR in which the user can select which service to backup. I think that's a better approach than to implement exclude.

Reply:
Sure, you can reverse the logic and use an `include` statement instead. I guess that's a toss-up. :)

Reply:
Oh sorry I mixed it up. We wanted to add the ability to choose individual services to _restore_. Here `--exclude` is actually more suitable, taking into account your other comment.